pkg/inventory/inventory-client.go

```diff
@@ -165,20 +226,147 @@ func (cic *ClusterInventoryClient) retrievePreviousInventoryObjects(current *obj
 	if err != nil {
 		return nil, err
 	}
-	cic.pastInventoryObjects = retrievedInventoryInfos
-	cic.retrievedInventoryObjects = true
-	klog.V(4).Infof("prune %d inventory objects found", len(cic.pastInventoryObjects))
-	return retrievedInventoryInfos, nil
+	var clusterInv *resource.Info
+	if len(retrievedInventoryInfos) == 1 {
+		clusterInv = retrievedInventoryInfos[0]
+	} else if len(retrievedInventoryInfos) > 1 {
+		clusterInv, err = cic.mergeClusterInventory(retrievedInventoryInfos)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return clusterInv, nil
+}
+
+// mergeClusterInventory merges the inventory of multiple inventory objects
+// into one inventory object, and applies it. Deletes the remaining unnecessary
+// inventory objects. There should be only one inventory object stored in the
+// cluster after this function. This special case should be very rare.
+//
+// TODO(seans3): Remove this code once we're certain no customers have multiple
+// inventory objects in their clusters.
+func (cic *ClusterInventoryClient) mergeClusterInventory(invInfos []*resource.Info) (*resource.Info, error) {
```

Comment:
How about doing this as part of kpt fix command so that we are transparent to the users about this migration? Agree that this is easier approach but just a thought.

Reply:
Let's keep this as is. We'd have to revert numerous function signatures to return a slice of inventory infos, instead of assuming just one inventory info is returned. This would be very awkward.